src/mpid/ch4/shm/posix/posix_progress.c

```diff
@@ -81,23 +78,28 @@ static int progress_recv(int blocking)
 
         /* note: setting is_local, is_async to 1, 1 */
         MPIDIG_global.target_msg_cbs[msg_hdr->handler_id] (msg_hdr->handler_id, am_hdr,
-                                                           NULL, in_total_data_sz, 1, 1, &rreq);
-
-        if (!rreq) {
-            MPIDI_POSIX_eager_recv_commit(&transaction);
-            goto fn_exit;
-        } else if (in_total_data_sz == payload_left) {
-            MPIDIG_recv_copy(p_data, rreq);
-            MPIDIG_REQUEST(rreq, req->target_cmpl_cb) (rreq);
-            MPIDI_POSIX_eager_recv_commit(&transaction);
-            MPIDI_POSIX_EAGER_RECV_COMPLETED_HOOK(rreq);
-            goto fn_exit;
-        } else {
-            /* prepare for asynchronous transfer */
-            MPIDIG_recv_setup(rreq);
-
-            MPIR_Assert(MPIDI_POSIX_global.active_rreq[transaction.src_local_rank] == NULL);
-            MPIDI_POSIX_global.active_rreq[transaction.src_local_rank] = rreq;
+                                                           NULL, payload_left, 1, 1, &rreq);
```

Comment:
I am thinking that you could distribute this callback into each cases below, then for `__HDR` and `__SHORT` cases, you could just call it with `is_async=0` and be done with it.